src/org/sosy_lab/common/collect/SkipList.java

```diff
@@ -0,0 +1,756 @@
+    // to recursive structure
+    private transient List<Node<T>> next;
+    private transient List<Node<T>> prev;
+    @Nullable private T value; // May be null, so be careful with serializing
+    private final int maxLvl;
+
+    private List<Integer> inBetweenCount;
+
+    private Node(@Nullable T pValue, int pLevel) {
+      value = pValue;
+      next = new ArrayList<>(Collections.nCopies(pLevel + 1, null));
+      prev = new ArrayList<>(Collections.nCopies(pLevel + 1, null));
+      inBetweenCount = new ArrayList<>(Collections.nCopies(pLevel + 1, 0));
+      maxLvl = pLevel;
+    }
+
+    @Nullable
+    Node<T> getNext(int pLevel) {
+      if (pLevel >= next.size()) {
+        return null;
+      } else {
+        return next.get(pLevel);
+      }
+    }
+
+    @Nullable
+    Node<T> getPrevious(int pLevel) {
+      if (pLevel >= prev.size()) {
+        return null;
+      } else {
+        return prev.get(pLevel);
+      }
+    }
+
+    @Nullable
+    T getValue() {
+      return value;
+    }
+
+    void setNext(@Nullable Node<T> pNext, int pLevel) {
+      next.set(pLevel, pNext);
+    }
+
+    void setPrevious(@Nullable Node<T> pPrev, int pLevel) {
+      prev.set(pLevel, pPrev);
+    }
+
+    void setInBetweenCount(int pNewValue, int pLevel) {
+      if (pLevel > maxLvl) {
+        // DO NOTHING
+      } else {
+        inBetweenCount.set(pLevel, pNewValue);
+      }
+    }
+
+    void increaseInBetweenCount(int pLevel) {
+      inBetweenCount.set(pLevel, inBetweenCount.get(pLevel) + 1);
+    }
+
+    int getInBetweenCount(int pLevel) {
+      if (pLevel > maxLvl) {
+        return 0;
+      } else {
+        return inBetweenCount.get(pLevel);
+      }
+    }
+
+    int getMaxLvl() {
+      return maxLvl;
+    }
+
+    private void readObject(ObjectInputStream pIn) throws IOException, ClassNotFoundException {
+      pIn.defaultReadObject();
+
+      next = new ArrayList<>(getMaxLvl());
+      prev = new ArrayList<>(getMaxLvl());
+    }
+  }
+
+  private static final int MAX_LEVEL = 31;
+  private static final double ONE_HALF_LOG = Math.log(0.5);
+  private static final Random randomGenerator = new Random();
+
+  private final Comparator<? super T> comparator;
+  private Node<T> head = createHead();
+  private Node<T> tail = head;
+  private int size = 0;
+
+  @SuppressWarnings("unchecked")
+  public SkipList(@Nullable Comparator<? super T> pComparator) {
+    if (pComparator == null) {
+      comparator = (Comparator<? super T>) Comparator.naturalOrder();
+    } else {
+      comparator = pComparator;
+    }
+  }
+
+  public SkipList(Collection<? extends T> pCollection) {
+    this();
+    boolean changed = addAll(pCollection);
+    assert changed;
+  }
+
+  public SkipList(SortedSet<T> pSortedSet) {
+    this(pSortedSet.comparator());
+    boolean changed = addAll(pSortedSet);
+    assert changed;
+  }
+
+  @SuppressWarnings("unchecked")
+  public SkipList() {
+    comparator = (Comparator<? super T>) Comparator.naturalOrder();
```

Comment:
The unchecked cast should be replaced with `Comparator.<? super T>naturalOrder()` (same above).

Reply:
Will this work?
`T` does not have to extend Comparable, but the type parameter of `Comparator.naturalOrder()` does have to.

There's actually a bigger, more general problem: if no comparator is given and `T` does not implement the Comparable-interface.
But with the two constructors creating `Comparator.naturalOrder()`, there is no way of telling because we do not have an object of Type `T` available.

Looking at the `TreeMap` implementation, they also do not handle this.

Reply:
This is another argument for factory methods, for those that use the natural order you can enfore `T extends Comparable<T>` via the type parameter of the method. Then you don't need the unchecked cast.